src/object_storage/s3.rs

```diff
@@ -1,39 +1,148 @@
+        id: String,
+        name: String,
+        access_key_id: String,
+        secret_access_key: String,
+        region: AwsRegion,
+        bucket_versioning_activated: bool,
+        bucket_ttl_in_seconds: Option<u32>,
+    ) -> Self {
         S3 {
             context,
             id,
             name,
-            access_key_id,
-            secret_access_key,
+            access_key_id: access_key_id.to_string(),
+            secret_access_key: secret_access_key.to_string(),
+            region: region.clone(),
+            bucket_versioning_activated,
+            bucket_ttl_in_seconds,
         }
     }
 
-    fn credentials_environment_variables(&self) -> Vec<(&str, &str)> {
-        vec![
-            (AWS_ACCESS_KEY_ID, self.access_key_id.as_str()),
-            (AWS_SECRET_ACCESS_KEY, self.secret_access_key.as_str()),
-        ]
+    fn get_credentials(&self) -> StaticProvider {
+        StaticProvider::new(self.access_key_id.clone(), self.secret_access_key.clone(), None, None)
+    }
+
+    fn get_s3_client(&self) -> S3Client {
+        let region = RusotoRegion::from_str(&self.region.to_aws_format())
+            .expect(format!("S3 region `{}` doesn't seems to be valid.", self.region.to_aws_format()).as_str());
+        let client = Client::new_with(
+            self.get_credentials(),
+            HttpClient::new().expect("unable to create new Http client"),
+        );
+
+        S3Client::new_with_client(client, region)
+    }
+
+    fn is_bucket_name_valid(bucket_name: &str) -> Result<(), Option<String>> {
+        if bucket_name.is_empty() {
+            return Err(Some("bucket name cannot be empty".to_string()));
+        }
+
+        Ok(())
+    }
+
+    pub fn bucket_exists(&self, bucket_name: &str) -> bool {
+        let s3_client = self.get_s3_client();
+
+        // Note: Using rusoto here for convenience, should be possible via CLI but would be way less convenient.
+        // Using retry here since there is a lag after bucket creation
+        retry::retry(Fixed::from_millis(1000).take(10), || {
+            block_on(s3_client.head_bucket(HeadBucketRequest {
+                bucket: bucket_name.to_string(),
+                expected_bucket_owner: None,
+            }))
+        })
+        .is_ok()
+    }
+
+    fn empty_bucket(&self, bucket_name: &str) -> Result<(), EngineError> {
+        if let Err(message) = S3::is_bucket_name_valid(bucket_name) {
+            let message = format!(
+                "While trying to empty S3 bucket, name `{}` is invalid: {}",
+                bucket_name,
+                message.unwrap_or_else(|| "unknown error".to_string())
+            );
+            return Err(self.engine_error(EngineErrorCause::Internal, message));
+        }
+
+        let s3_client = self.get_s3_client();
+
+        // make sure to delete all bucket content before trying to delete the bucket
+        let objects_to_be_deleted = match block_on(s3_client.list_objects(ListObjectsRequest {
```

Comment:
We've seen with Loki how painful it is to wait for a long time for some s3 buckets. May be we have another solution to speed this up?

Reply:
Unfortunately there is no such things via Rusoto, no option to delete a non empty bucket ... I can switch back to CLI but well ... we won't be able to remove it then.

Reply:
Let's keep it this way